pinot-common/src/main/java/org/apache/pinot/common/request/context/RequestContextUtils.java

```diff
@@ -411,4 +415,69 @@ public static FilterContext getFilter(FilterQueryTree node) {
         throw new IllegalStateException();
     }
   }
+
+  public static FilterContext overrideFilterWithExpressionOverrideHints(FilterContext filter, IndexSegment indexSegment,
+      Map<ExpressionContext, ExpressionContext> getExpressionOverrideHints) {
+    if (filter.getChildren() != null) {
+      // AND, OR, NOT
+      List<FilterContext> newChildren = filter.getChildren().stream().map(
+              filterContext -> overrideFilterWithExpressionOverrideHints(filterContext, indexSegment,
+                  getExpressionOverrideHints))
+          .collect(Collectors.toList());
+      return new FilterContext(filter.getType(), newChildren, null);
+    }
+    // PREDICATE
+    Predicate predicate = filter.getPredicate();
+    ExpressionContext newPredicateLhs =
+        overrideWithExpressionHints(indexSegment, getExpressionOverrideHints, predicate.getLhs());
+    Predicate newPredicate = getNewPredicate(predicate, newPredicateLhs);
+    return new FilterContext(filter.getType(), null, newPredicate);
+  }
+
+  public static Predicate getNewPredicate(Predicate predicate, ExpressionContext lhs) {
```

Comment:
To save garbage, we can add `setLhs()` to all predicates. Consider adding a `BasePredicate` with `setLhs()` and `getLhs()` and we don't need the switch